# update bim SNP id to dbSNP 146

```{r, eval=FALSE}

dbpath=/statgen/meiwen/CHD/mw/08imputation/download
snptracker=/statgen/meiwen/CHD/tools/snptracker/snptracker.jar
cd /statgen/meiwen/CHD/mw/08imputation/download
mv All_20151104.vcf.gz All_20151104.dbSNP146.vcf.gz
gunzip -c All_20151104.dbSNP146.vcf.gz > All_20151104.dbSNP146.vcf &

cd ..

awk '{if($1==23){$1="X"}; OFS="\t"; print $0}' CHD_AllQC_rsid.bim > CHD_AllQC_rsid.SEXasX.bim
java -Xmx4g -jar ${snptracker} \
   --bim-file CHD_AllQC_rsid.SEXasX.bim --by-pos --ref hg19 \
   --out CHD_AllQC_rsid.SEXasX \
   --merge-file ${dbpath}/RsMergeArch.bcp.gz --coor-file ${dbpath}/b146_SNPChrPosOnRef_105.bcp.gz --hist-file ${dbpath}/SNPHistory.bcp.gz 2>&1 > snptrack.log

awk '$1==0' CHD_AllQC_rsid.SEXasX.result.bim | cut -f 2 | sed -e 's/rsAX-/AX-/g' > CHD_AllQC_rsid.SEXasX.result.unmapped.list
plink --bfile CHD_AllQC_rsid --exclude CHD_AllQC_rsid.SEXasX.result.unmapped.list --make-bed --out CHD_AllQC_rsid.tmp
awk '$1!=0' CHD_AllQC_rsid.SEXasX.result.bim > CHD_AllQC_rsid.tmp.bim

/statgen/meiwen/CHD/tools/snpflip/bin/snpflip -b CHD_AllQC_rsid.tmp.bim -f $dbpath/human_g1k_v37.fasta -o CHD_AllQC_rsid.tmp.snpflip_output 

plink --bfile CHD_AllQC_rsid.tmp --flip CHD_AllQC_rsid.tmp.snpflip_output.reverse --make-bed --out CHD_AllQC_rsid.fliped
awk 'NR>1{OFS="\t"; print $3,$7}' CHD_AllQC_rsid.tmp.snpflip_output.annotated_bim > CHD_AllQC_rsid.fliped.ref.allele

plink --bfile CHD_AllQC_rsid.fliped --a2-allele CHD_AllQC_rsid.fliped.ref.allele 2 1 --make-bed --out CHD_AllQC_rsid.fliped.ref
grep "Impossible A2" CHD_AllQC_rsid.fliped.ref.log | cut -f 8 -d ' ' | sed -e 's/\.//g' > CHD_AllQC_rsid.fliped.ref.exclude
plink --bfile CHD_AllQC_rsid.fliped --exclude CHD_AllQC_rsid.fliped.ref.exclude --a2-allele CHD_AllQC_rsid.fliped.ref.allele 2 1 --make-bed --out CHD_AllQC_rsid.fliped.ref.1
plink --bfile CHD_AllQC_rsid.fliped.ref.1 --keep-allele-order --freq --out CHD_AllQC_rsid.fliped.ref.1
plink --bfile CHD_AllQC_rsid.fliped --keep-allele-order --freq --out CHD_AllQC_rsid.fliped
head CHD_AllQC_rsid.fliped.ref.1.frq CHD_AllQC_rsid.fliped.frq

awk 'OFS="\t"{if($1==23){$1="X"}; print $0}' CHD_AllQC_rsid.fliped.ref.1.bim > CHD_AllQC_rsid.fliped.ref.1.bim.X
mv CHD_AllQC_rsid.fliped.ref.1.bim.X CHD_AllQC_rsid.fliped.ref.1.bim

extractbyquery.pl CHD_AllQC_rsid.fliped.ref.1.bim 2 download/All_20151104.dbSNP146.vcf 3 | awk 'OFS="\t"{print $1,$2,$3,$4,$5}' > CHD_AllQC_rsid.fliped.ref.1.dbSNP146
cut -f 3 CHD_AllQC_rsid.fliped.ref.1.dbSNP146 > CHD_AllQC_rsid.fliped.ref.1.dbSNP146.snplist
plink --bfile CHD_AllQC_rsid.fliped.ref.1 --keep-allele-order --extract CHD_AllQC_rsid.fliped.ref.1.dbSNP146.snplist --make-bed --out CHD_AllQC_rsid.fliped.ref.2

paste CHD_AllQC_rsid.fliped.ref.1.dbSNP146 CHD_AllQC_rsid.fliped.ref.2.bim | awk '$3==$7{OFS="\t";
      if(($10=="-" && (length($4)==2 || length($5)==2) ) || ($11=="-" && (length($4)==2 || length($5)==2))){$9=$2;$10=$5;$11=$4;} 
      if(length($10)==1 || length($11)==1){ print $0; } }' | cut -f 6-11 | grep -P -v ",|-"  > CHD_AllQC_rsid.fliped.ref.2.dbSNP146.biallelic.bim

wc -l  CHD_AllQC_rsid.fliped.ref.2.dbSNP146.biallelic.bim CHD_AllQC_rsid.fliped.ref.2.bim
cut -f 2 CHD_AllQC_rsid.fliped.ref.2.dbSNP146.biallelic.bim > CHD_AllQC_rsid.fliped.ref.2.dbSNP146.biallelic.snplist

plink --bfile CHD_AllQC_rsid.fliped.ref.2 --keep-allele-order --extract CHD_AllQC_rsid.fliped.ref.2.dbSNP146.biallelic.snplist --make-bed --out CHD_AllQC_rsid.fliped.ref.3
wc -l CHD_AllQC_rsid.fliped.ref.2.dbSNP146.biallelic.bim  CHD_AllQC_rsid.fliped.ref.3.bim
head CHD_AllQC_rsid.fliped.ref.2.dbSNP146.biallelic.bim  CHD_AllQC_rsid.fliped.ref.3.bim
cp CHD_AllQC_rsid.fliped.ref.2.dbSNP146.biallelic.bim  CHD_AllQC_rsid.fliped.ref.3.bim

plink --bfile CHD_AllQC_rsid.fliped.ref.3 --keep-allele-order --recode vcf-iid  --out CHD_AllQC_rsid.fliped.ref
head -n 70 CHD_AllQC_rsid.fliped.ref.vcf | cut -f 1-6
head CHD_AllQC_rsid.fliped.ref.allele

grep -v '#' CHD_AllQC_rsid.fliped.ref.vcf | cut -f 4  | sort | uniq -c
grep -v '#' CHD_AllQC_rsid.fliped.ref.vcf | cut -f 5  | sort | uniq -c
# grep '#' CHD_AllQC_rsid.fliped.ref.vcf 

cp CHD_AllQC_rsid.fliped.ref.vcf impute.vcf
bgzip impute.vcf
LD_LIBRARY_PATH=~/usr/lib/ tabix impute.vcf.gz

#wget https://imputation.sanger.ac.uk/www/plink2ensembl.txt
bcftools annotate -Oz --rename-chrs plink2ensembl.txt impute.vcf.gz > imputechrensembl.vcf.gz
LD_LIBRARY_PATH=~/usr/lib/ tabix imputechrensembl.vcf.gz
zcat imputechrensembl.vcf.gz | head -n 28

sed -e 's/>23/>X/' human_g1k_v37.fasta > human_g1k_v37.ensembl.fasta
awk 'BEGIN{b=1;}{if($0~/>Y/){b=0} if(b==1){print $0}}' human_g1k_v37.ensembl.fasta > human_g1k_v37.ensembl_1-X.fasta
grep ">" human_g1k_v37.ensembl_1-X.fasta

grep 5739516 *.bim | head
grep 5739516 download/*.vcf | head 

#bcftools norm -c wxs -f ${dbpath}/human_g1k_v37.ensembl_1-X.fasta imputechrensembl.vcf.gz -O z > imputechrensembl.norm.vcf.gz
bcftools norm -c w -f ${dbpath}/human_g1k_v37.ensembl_1-X.fasta imputechrensembl.vcf.gz -O z > imputechrensembl.norm.vcf.gz
Lines   total/split/realigned/skipped:  429257/0/0/0

LD_LIBRARY_PATH=~/usr/lib/ tabix imputechrensembl.norm.vcf.gz

cut -d ' ' -f 5 CHD_AllQC_rsid.fliped.ref.1.fam | sort | uniq -c
awk 'OFS="\t"{if($5==2){$5="F"}if($5==1){$5="M"} print $2,$5}' CHD_AllQC_rsid.fliped.ref.1.fam > samples.txt

```

## globus

```{r,eval=F}

which virtualenv || pip install virtualenv
LD_LIBRARY_PATH=/home/meiwen/anaconda2/lib/ virtualenv "$HOME/.globus-cli-virtualenv"
echo "export LD_LIBRARY_PATH=/home/meiwen/anaconda2/lib/" >> $HOME/.globus-cli-virtualenv/bin/activate
source "$HOME/.globus-cli-virtualenv/bin/activate"
pip install globus-cli
deactivate
export PATH="$PATH:$HOME/.globus-cli-virtualenv/bin"

globus login
globus get-identities 'zhaochen.tcm@gmail.com'

globus endpoint create --personal globusimputepe2
Message:     Endpoint created successfully
Endpoint ID: 6a17ce90-20fc-11e7-bc3c-22000b9a448b
Setup Key:   f1f8a811-d5c5-454c-a520-fad51a35b07e

./globusconnectpersonal -setup f1f8a811-d5c5-454c-a520-fad51a35b07e
### fail to connect, cp to HM server

cd /home/chenzhao/globusshare/
rsync -avzP sichao@202.127.22.38:~/samples.txt ./globusshare/
rsync -avzP sichao@202.127.22.38:~/imputechrensembl.norm.v* ./globusshare/

# build a test data
zcat imputechrensembl.norm.vcf.gz | awk 'NR%100>0 || $1~/#/' | bgzip -c /dev/stdin > imputechrensembl.test.norm.vcf.gz
tabix imputechrensembl.test.norm.vcf.gz

# upload 

```
## impute

impute by:

HRC panel and UK10K+1000g panel;
   HRC panel for 
       
       all data
       
       test data

   UK10k+1000g panel for 
       
       all data
           stand pipelien
       test dat, 
           stand pipelien
           shapeit2 pipeline

## download 

```{r, eval=F}

835f17e9-6907-4fa5-a918-0ae94073fa28

git clone https://github.com/rofl0r/proxychains-ng.git
configure --prefix=~/usr --sysconfdir=~/etc
make -j 8 and make install
make install-config

@ pcl167

cd /home/meiwen/statgen/CHD/mw/08imputation
proxychains4 -f ~/etc/proxychains.conf ./globusconnectpersonal-2.3.3/globusconnect

## firewall failed

download in HM server

```

## impute QC

```{r,eval=F}

# test autosome
cd /home/chenzhao/globusshare/
zcat imputechrensembl.norm.vcf.gz | awk 'NR%100==0 || $1~/#/' > imputechrensembl.test.norm.true.vcf

cd /home/chenzhao/globusshare/imputecathyQC.vcfs
cd /home/chenzhao/globusshare/imputecathy10k100gQC.vcfs

rm imputechrensembl.test.norm.pred.vcf
for i in {1..22}; do
zcat $i.vcf.gz | extractbyquery.pl ../imputechrensembl.test.norm.true.vcf 3 /dev/stdin 3 | grep -v '#' >> imputechrensembl.test.norm.pred.vcf ; 
done;

extractbyquery.pl imputechrensembl.test.norm.pred.vcf 3 ../imputechrensembl.test.norm.true.vcf 3 | grep -v '#' > imputechrensembl.test.norm.true.vcf

cd /home/chenzhao/globusshare/imputecathy10k100gQCshapeit2.vcfs
rm imputechrensembl.test.norm.pred.vcf
for i in {1..22}; do
zcat $i.pbwt_reference_impute.vcf.gz | extractbyquery.pl ../imputechrensembl.test.norm.true.vcf 3 /dev/stdin 3 | grep -v '#' >> imputechrensembl.test.norm.pred.vcf ;
done

extractbyquery.pl imputechrensembl.test.norm.pred.vcf 3 ../imputechrensembl.test.norm.true.vcf 3 | grep -v '#' > imputechrensembl.test.norm.true.vcf


R


plot_acccy <- function(path="/home/chenzhao/globusshare/imputecathy10k100gQC.vcfs",pipeline="EAGLE2+PBWT"){
   setwd(path) 
   itrue <- read.table("imputechrensembl.test.norm.true.vcf",stringsAsFactors=F,header=F)
   ipred <- read.table("imputechrensembl.test.norm.pred.vcf",stringsAsFactors=F,header=F)

   iinfo <- as.numeric(gsub(".*INFO=(.*)","\\1",ipred[[8]]))
   GTidx <- c("0/0","0/1","0/1","1/1")
   names(GTidx) <- c("0|0","0|1","1|0","1|1")
   if(pipeline=="SHAPEIT2+PBWT"){
      names(GTidx) <- c("0/0","0/1","1/0","1/1")
   }
   GPidx <- c(NA,2,1,0)
   names(GPidx) <- c("./.","0/0","0/1","1/1")
   ipred_GT <- t(apply(ipred[,-(1:9)],1,function(x){
      GTidx[gsub(":.*","",x)]
   }))
   iaccy_GT <- sapply(1:nrow(ipred_GT),function(i){
      x <- as.character(itrue[i,-c(1:9)])
      y <- ipred_GT[i,]
      sum(x==y)/length(x)
   })
   ipred_GP <- t(apply(ipred[,-(1:9)],1,function(x){
      sapply(strsplit(gsub(".*:","",x),","),function(y){y<-as.numeric(y);y[1]*2+y[2]})
   }))
   itrue_GP <- t(apply(itrue[,-(1:9)],1,function(x){
      GPidx[x]
   }))
   iaccy_meandif_GP <- apply(itrue_GP-ipred_GP,1,function(x){mean(abs(x),na.rm=T)})
   plot(iinfo,iaccy_GT)
   plot(iinfo,iaccy_meandif_GP)
}

par(mfrow=c(3,2))

plot_acccy("/home/chenzhao/globusshare/imputecathy10k100gQC.vcfs",pipeline="EAGLE2+PBWT")
plot_acccy("/home/chenzhao/globusshare/imputecathyQC.vcfs",pipeline="EAGLE2+PBWT")
plot_acccy("/home/chenzhao/globusshare/imputecathy10k100gQCshapeit2.vcfs",pipeline="SHAPEIT2+PBWT")



```

## vcf filter

```{r,eval=F}
git clone --recursive https://github.com/ekg/vcflib.git

cd /home/chenzhao/globusshare/imputecathy.vcfs

tabix 22.vcf.gz
zcat 22.vcf.gz | cut -f 1-10 | vcffilter -f ' INFO > 0.1 ' | wc -l

zcat 22.vcf.gz | cut -f 1-10 | vcffilter -f ' (RefPanelAF > 0.005 & RefPanelAF < 0.995) & INFO > 0.1 ' | wc -l
zcat 22.vcf.gz | cut -f 1-10 | vcffilter -f ' (( RefPanelAF > 0.005 ) & ( RefPanelAF < 0.995 )) & ( INFO > 0.1 ) '  | wc -l



```

## snptest 

```{r,eval=F}

zcat 22.vcf.gz | vcffilter -f ' (( RefPanelAF > 0.005 ) & ( RefPanelAF < 0.995 )) & ( INFO > 0.1 ) ' | head -n 10000| bgzip -c /dev/stdin > test.snptest.vcf.gz

echo 'ID_1 ID_2 missing sex chd mds_1 mds_2 mds_3 mds_4 mds_5 mds_6 mds_7 mds_8 mds_9 mds_10' > sampleinfo.snptest.sample
echo '0 0 0 D B C C C C C C C C C C' >> sampleinfo.snptest.sample

awk 'NR>1' CHD_iQC_lQC02_PCA_lQC_mds.mds | tr -s '\t' | tr -s ' ' | tr '\t' ' ' | sed -e 's/^ //' |  cut -d ' '  -f 4- > sampleinfo.snptest.mds
cut -d ' ' -f 1,2,3,5,6  CHD_AllQC_rsid.fam | awk '{$4=2-$4; $5=$5-1; print $0}'  > sampleinfo.snptest.pheno
paste -d ' ' sampleinfo.snptest.pheno sampleinfo.snptest.mds >> sampleinfo.snptest.sample

snptest_v2.5.4-beta2 -data test.snptest.vcf.gz sampleinfo.snptest.sample -o test.snptest.out  -genotype_field GP -frequentist 1 -method expected -pheno chd -hwe -log test.snptest.log

grep alternate_ids test.snptest.out | tr ' ' '\n' | awk '{print NR,$0}'

```

